### ISSUE TITLE
Git Extensions keeps handles open to repo directory, preventing it from being deleted

### ISSUE THREAD

**Issue by loop-evgeny** [gitextensions/gitextensions#4225]
**Do you want to request a *feature* or report a *bug*?** Bug

**What is the current behavior?** When I had a repository open in GE and then open a different repository (by selecting a dropdown item from the toolbar) GE still keeps file handles on the repository directory itself and its .git sub-directory (I can see these in Process Hacker). This prevents the directory from being deleted.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.** Open GE, open repository 1, change to repository 2.

**What is the expected behavior?** GE closes all handles to the previously opened repository, so it can be deleted.

**Environment you encounter the issue:**

- GitExtensions version: 2.50.02
- GIT version:  2.15.0.windows.1
- OS version:  Windows 7 N x64 SP1

**Did this work in previous version of GitExtensions (which)?** Don't know

**Comment by loop-evgeny** [comment]
Still an issue in 3.1.1. This is quite annoying.

**Comment by pmiossec** [comment]
>Still an issue in 3.1.1. This is quite annoying.

This is the typical kind of bug that annoy very few people (that have a specific workflow).
GitExtensions evolve with contributions so, if you want to see it fixed, you are the one the more able to fix it.
Please, contribute!

**Comment by loop-evgeny** [comment]
I'm a bit surprised more people don't run into this.

I had a go at debugging it. It seems to be due to the FileSystemWatchers created by various FileSettingsCache objects, which are never disposed. Even if I add a finalizer to FileSettingsCache it never gets called when changing between repositories, either (even with `GC.Collect()`), so something is still holding a reference to it - I haven't figured out what.

**Comment by GintasS** [comment]
I think this is still relevant. Tested it and couldn't delete a directory while GE was open.

**Comment by RussKie** [comment]
PRs with a fix a welcome

**Comment by SAPHemlock** [comment]
Still reproducible at master. Posting the root cause since it is concrete, plus why a drive-by fix would be a bad idea.

**The handle.** `FileSettingsCache` owns a `FileSystemWatcher` that is armed at construction and never disposed:

```csharp
private readonly FileSystemWatcher _fileWatcher = new();
...
_fileWatcher.Path = dir;
_fileWatcher.Filter = Path.GetFileName(SettingsFilePath);
_fileWatcher.EnableRaisingEvents = true;   // holds a handle on `dir`
```

`DistributedSettings` creates two of them per repository - on `module.GitCommonDirectory` and on `module.WorkingDir` - which is exactly the pair reported here (the repo directory and its `.git`). Since `7c2fe0a0c` ("Always watch repo-local settings", Refs #11781) the watcher is armed even when `GitExtensions.settings` does not exist, so every repository ever opened contributes two directory handles.

They are never released because nothing owns them: `GitModule`, `GitUICommands`, `SettingsContainer` and `DistributedSettings` are not `IDisposable`, a repo switch just replaces the object graph (`FormBrowse.SetGitModule` -> `UICommands.WithGitModule(...)` -> `new GitUICommands(...)`), and `FileSettingsCache.Dispose` - which *does* correctly dispose the watcher - has no caller anywhere in `src/`. GC does not save it either: an armed `FileSystemWatcher` keeps its `SafeFileHandle` alive inside the pending `ReadDirectoryChangesW` state, which is why the finalizer experiment in this thread never ran.

For completeness, the other candidates are fine: `GitStatusMonitor`'s watchers are re-pointed (setting `FileSystemWatcher.Path` restarts an enabled watcher, closing the old handle) and disposed; `IndexWatcher` clears and disposes; processes use `using IProcess`; `GitFileStreamGetter` returns `MemoryStream`. The console tab is best-effort by design (it is told to `cd` on switch). One genuinely separate handle: `GitModule` calls `Directory.SetCurrentDirectory(WorkingDir)` in a few places and never restores it, but that only bites after conflict resolution, so it does not explain the plain repro.

**Why I am not opening a PR for it.** The caches are shared through an unowned weak cache:

```csharp
public static T FromCache<T>(string settingsFilePath, Lazy<T> createSettingsCache)
    where T : FileSettingsCache
    => WeakRefCache.Default.Get(settingsFilePath + ":" + typeof(T).FullName, createSettingsCache);
```

The same instance is handed to submodule/superproject modules, other browse windows, `FormCommit` and plugins, with no refcount, and `WeakRefCache` will keep handing out an instance that is alive-but-disposed. Disposing at repo-switch time would therefore set `_saveTimer` to null for someone else still using that repo, so `StartSaveTimer` silently drops their setting writes - losing user settings is far worse, and far less visible, than a stuck directory handle. A correct fix needs either ownership/refcounting in `WeakRefCache` plus a disposable `GitModule`, or dropping the watcher in favour of the existing mtime polling - and the latter would revert `f31d19fd8` and the intent of `7c2fe0a0c`.

There is one narrow, safe sub-fix that does **not** address this repro: `CommonLogic` creates three caches with `useSharedCache: false` (exclusively owned, `autoSave: false`) per Settings-dialog open and never disposes them, leaking two more repo-directory handles each time. Making `CommonLogic` disposable would fix that leak alone. Happy to submit it if that is wanted, but it needs care because `EffectiveSettings` shares the Local/Distributed caches, so the set has to be deduplicated.

**Comment by SAPHemlock** [comment]
I need to correct my earlier comment on this issue. I attributed the locked directory to the undisposed `FileSystemWatcher`s in `FileSettingsCache`, following @emz00's 2019 debugging. I have now measured it, and **that mechanism does not hold the directory on .NET 10.**

```
.NET 10.0.12
1) watcher ARMED, never disposed -> Directory.Delete(dir, recursive: true) succeeded: True
2) watcher DISPOSED              -> Directory.Delete(dir, recursive: true) succeeded: True
```

`FileSystemWatcher` opens its directory with `FILE_SHARE_DELETE`, so the handle is visible in Process Hacker but does not block removal. A visible handle is not the same as a lock, and I conflated the two.

### What does hold it

The process current directory, which GE sets and never restores:

```
1) CWD elsewhere       -> delete succeeded: True
2) CWD set to the dir  -> delete succeeded: False
      IOException: The process cannot access the file '...' because it is being used by another process.
3) CWD restored, retry -> delete succeeded: True
```

That error text is exactly the symptom this issue is about. The call sites are:

- `GitModule.cs:526` — `HandleConflictSelectSide`
- `GitModule.cs:552` — `HandleConflictsSaveSide`
- `GitModule.cs:658` — `CheckoutConflictedFiles`
- `FormResolveConflicts.cs:175` and `:1305`

all of them `Directory.SetCurrentDirectory(WorkingDir)` with nothing setting it back.

But note what that means for this report: **every one of those is on a conflict-resolution path.** @emz00's repro is "open repo 1, switch to repo 2" with no conflicts involved, so this does not explain it either. On current master I cannot identify a mechanism that would reproduce the reported steps, and I could not reproduce them myself.

So the honest state of this issue is: the 2017 diagnosis was for .NET Framework, the mechanism everyone (including me) has been pointing at has since stopped having that effect, and the one confirmed way GE still pins a repository directory is reachable only after resolving conflicts. @GintasS, you reproduced it in 2021 — does it still happen for you on a current build, and if so does it need a conflict resolution first?

### What I have sent, and what I have not

#13325 disposes the settings caches `CommonLogic` owns exclusively. Those watchers really are leaked — three more per settings-dialog open, never released — so it is worth fixing as a resource leak. Given the measurement above I have been explicit in that PR that **it does not close this issue.**

I have deliberately not touched the shared caches from `GitModule.EffectiveSettings`, which is what would be needed for a repo switch to release anything. They are handed out by `WeakRefCache` with no ownership or refcount, to submodules, other browse windows and plugins; disposing one nulls its `_saveTimer` and another holder's next write is silently dropped. That needs ownership in `WeakRefCache` first, and I would not send it as a drive-by.

The `SetCurrentDirectory` calls look genuinely fixable and narrow — restore the previous directory, or better, stop relying on the process CWD for those `checkout-index` invocations. Happy to do that as its own PR if a maintainer thinks it is worth it, though it fixes a different scenario than the one filed here.

*Measured with [Claude Code](https://claude.com/claude-code).*
